kinto/core/events.py

```diff
@@ -116,31 +116,50 @@ def on_new_request(event):
                      under=pyramid.tweens.EXCVIEW)
 
 
-def get_resource_events(request, after_commit=False):
+def drain_resource_events(request):
     """
     Request helper to return the list of events triggered on resources.
     The list is sorted chronologically (see OrderedDict)
     """
     by_resource = request.bound_data.get('resource_events', {})
-    events = []
-    for (action, payload, impacted, request) in by_resource.values():
-        if after_commit:
-            if action == ACTIONS.READ:
-                event_cls = AfterResourceRead
-            else:
-                event_cls = AfterResourceChanged
+    afterwards = request.bound_data.setdefault('after_resource_events', OrderedDict())
+
+    while by_resource:
+        key = next(iter(by_resource.keys()))
+        event = by_resource.pop(key)
+        (action, payload, impacted, request) = event
+
+        after_event = afterwards.get(key, None)
+        if not after_event:
+            afterwards[key] = event
         else:
-            if action == ACTIONS.READ:
-                event_cls = ResourceRead
-            else:
-                event_cls = ResourceChanged
-        event = event_cls(payload, impacted, request)
-        events.append(event)
-    return events
+            after_impacted = afterwards[key][2]
+            after_impacted.extend(impacted)
```

Comment:
Basically, I changed `resource_events` to be mutated (drained), but this poses a problem because we would no longer be able to fire `AfterResourceChanged` and `AfterResourceRead`, so now I keep a second explicit dictionary of those after they've been drained from the first one.